### PR TITLE
Validate custom time.Duration types

### PR DESCRIPTION
**Pull request by pgporada**
We can now validate custom `time.Duration` types defined as `type MyDuration time.Duration` with all of the baked in tags such as `gt`, `gte`, `eq`, `lt`, `lte,` etc.

In an application, the caller will need to define their custom duration, instantiate a new validator, register the custom duration type, then configure the validation tags where applicable. What that looks like in practice is
```
package example

// Sometimes we need a custom type because we have application specific methods
type MyDuration time.Duration

func (m *MyDuration) UnmarshalJson() ...
func (m *MyDuration) MarshalJson() ...
```

```
package config

type MyConfig struct {
  TLS TLSConfig
  ConnectMaxIdleTime example.MyDuration `validate:"omitempty,lte=60s"`
  ConnectionLifetime example.MyDuration `validate:"required,gte=5s"`
}

func ValidateConfig(mc MyConfig, in io,Reader) {
	// Initialize the validator and load any custom tags.
	validate := validator.New()
	for tag, v := range cv.Validators {
		err := validate.RegisterValidation(tag, v)
		if err != nil {
			return err
		}
	}

	// Register example.MyDuration so the validator understands it and map it to an existing type.
	validate.RegisterCustomType(time.Duration(0), example.MyDuration(0))

       err := validate.Struct(mc)
}
```

### CODE REVIEW

**Comment by pgporada** [comment]
Thank you @aaomidi for helping me figure this out.

File: util.go [hunk 1]
```diff
@@ -240,15 +242,21 @@ func asIntFromTimeDuration(param string) int64 {
 	return int64(d)
 }
 
+// GlobalTimeDurationTypes holds specific application registered custom
+// time.Duration types so their values can be properly parsed as time.Duration.
+var GlobalTimeDurationTypes = []reflect.Type{timeDurationType}
```
Comment (aarongable) [inline util.go]: AFACT no need for this to be exported:
```suggestion
var globalTimeDurationTypes = []reflect.Type{timeDurationType}
```

File: util.go [hunk 2]
```diff
@@ -240,15 +242,21 @@ func asIntFromTimeDuration(param string) int64 {
...
 // asIntFromType calls the proper function to parse param as int64,
 // given a field's Type t.
 func asIntFromType(t reflect.Type, param string) int64 {
-	switch t {
-	case timeDurationType:
+	for _, customDurationType := range GlobalTimeDurationTypes {
+		if t != customDurationType {
+			continue
+		}
 		return asIntFromTimeDuration(param)
```
Comment (aarongable) [inline util.go]: I'd reverse this logic:
```suggestion
		if t == customDurationType {
			return asIntFromTimeDuration(param)
		}
```

File: validator_instance.go [hunk 3]
```diff
@@ -222,6 +222,17 @@ func (v *Validate) RegisterValidationCtx(tag string, fn FuncCtx, callValidationE
...
 
+// RegisterDurationType is used to register a custom time.Duration type in your
+// application such as the following example. Reasons for implementing a custom
+// time.Duration would be for application specific methods. Note that the
+// validator will handle an aliased time.Duration without the use of this
+// method.
+//
+// Example: type MyDuration time.Duration
+func (v *Validate) RegisterDurationType(t any) {
+	GlobalTimeDurationTypes = append(GlobalTimeDurationTypes, reflect.TypeOf(t))
```
Comment (aaomidi) [inline validator_instance.go]: I do not think this should mutate a global state. This should probably belong to the specific validator instance.

File: validator_instance.go [hunk 4]
```diff
@@ -222,6 +222,17 @@ func (v *Validate) RegisterValidationCtx(tag string, fn FuncCtx, callValidationE
 	return v.registerValidation(tag, fn, false, nilCheckable)
 }
 
+// RegisterDurationType is used to register a custom time.Duration type in your
```
Comment (aarongable) [inline validator_instance.go]: nit: this should move below `registerValidation`, since that function (currently directly below this one) is a helper for `RegisterValidation` and `RegisterValidationCtx` (both directly above this one).

File: validator_instance.go (hunk 3 above)
Comment (aarongable) [reply validator_instance.go]: Oh good catch, I hadn't looked closely enough to realize that the other RegisterFoo functions were operating on their `*Validate` receiver. Agreed.

File: baked_in.go [hunk 5]
```diff
@@ -229,47 +229,47 @@ func parseOneOfParam2(s string) []string {
...
 	return false
 }
 
 // isUnique is the validation function for validating if each array|slice|map value is unique
-func isUnique(fl FieldLevel) bool {
+func isUnique(v *Validate, fl FieldLevel) bool {
 	field := fl.Field()
 	param := fl.Param()
-	v := reflect.ValueOf(struct{}{})
+	val := reflect.ValueOf(struct{}{})
```
Comment (beautifulentropy) [inline baked_in.go:272]: Is this rename necessary?

```suggestion
	v := reflect.ValueOf(struct{}{})
```

File: baked_in.go [hunk 6]
```diff
@@ -315,56 +315,56 @@ func isUnique(fl FieldLevel) bool {
 }
 
 // isMAC is the validation function for validating if the field's value is a valid MAC address.
-func isMAC(fl FieldLevel) bool {
+func isMAC(v *Validate, fl FieldLevel) bool {
```
Comment (beautifulentropy) [inline baked_in.go:318]: Here and below, `v *Validate` param is being added but never used. I'm not understanding why this is the case.

**Review by beautifulentropy — CHANGES_REQUESTED** [review]

File: baked_in.go (hunk 5 above)
Comment (pgporada) [reply baked_in.go:272]: Yes because of the new input parameter `v *Validate`. That parameter is needed to meet the function signature defined [here](https://github.com/letsencrypt/validator/pull/4/files#diff-6eda4809e29e9c01c780a30c46947d23b10606dbd948fd00dc5376030cf897c6R22). Without the rename I get an a "no new variables on left side" error.

File: baked_in.go (hunk 6 above)
Comment (pgporada) [reply baked_in.go:318]: All [bakedInValidators](https://github.com/letsencrypt/validator/pull/4/files#diff-6eda4809e29e9c01c780a30c46947d23b10606dbd948fd00dc5376030cf897c6L64-R73) have to match the `Func` function signature defined [here](https://github.com/letsencrypt/validator/pull/4/files#diff-6eda4809e29e9c01c780a30c46947d23b10606dbd948fd00dc5376030cf897c6R22). Should this parameter be `func isMAC(_ *Validate, fl FieldLevel) bool` instead since I don't use `v`?

Comment (aarongable) [reply baked_in.go:318]: Having not taken the time to fully understand the whole validator package myself, it _appears_ that there's something very weird going on where some things happen at the `*Validate` instance level, and others happen at the global level. This existing inconsistency is leading this PR to become very large while attempting to accomplish a seemingly-straightforward goal.

I don't think that this PR is the right place to do a massive refactoring of this library to allow everything to happen at the instance level. If we want to do that (and maybe we do!) it should be as a separate PR, where we can consider multiple different ways of making that happen (e.g. making lots of current top-level helper functions instead be methods on that type, which seems cleaner than having those methods take an instance as their first argument).

**Comment by pgporada** [comment]
Closing in favor of https://github.com/letsencrypt/boulder/pull/7564
